libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -238,64 +265,122 @@ string IRGenerator::generateInitialAssignment(VariableDeclaration const& _varDec
-		out << generator.code();
-
-		out << "// End state variable initialization for contract \"" << contract->name() << "\".\n";
+		FunctionDefinition const* baseConstructor = dynamic_cast<ContractDefinition const*>(
+			base->name().annotation().referencedDeclaration
+		)->constructor();
+
+		if (baseConstructor && base->arguments() && !base->arguments()->empty())
+			baseConstructorArguments.emplace_back(
+				dynamic_cast<ContractDefinition const*>(baseConstructor->scope()),
+				*base->arguments()
+			);
 	}
 
-	if (constructor)
-	{
-		ABIFunctions abiFunctions(m_evmVersion, m_context.revertStrings(), m_context.functionCollector());
-		unsigned paramVars = make_shared<TupleType>(constructor->functionType(false)->parameterTypes())->sizeOnStack();
+	IRGeneratorForStatements generator{m_context, m_utils};
+	for (pair<ContractDefinition const*, std::vector<ASTPointer<Expression>>> const& baseArgs: baseConstructorArguments)
+		if (baseArgs.first->constructor() && !baseArgs.second.empty())
+		{
+			string params;
+			for (size_t i = 0; i < baseArgs.second.size(); ++i)
+				params += (params.empty() ? "" : ", ") + generator.evaluateExpression(
+						*(baseArgs.second.at(i)),
+						*(baseArgs.first->constructor()->parameters()[i]->type())
+				).commaSeparatedList();
+			constructorParams[baseArgs.first] = params;
+		}
 
-		Whiskers t(R"X(
-			let programSize := datasize("<object>")
-			let argSize := sub(codesize(), programSize)
+	return {generator.code(), constructorParams};
+}
 
-			let memoryDataOffset := <allocate>(argSize)
-			codecopy(memoryDataOffset, programSize, argSize)
+string IRGenerator::initStateVariables(ContractDefinition const& _contract)
+{
+	IRGeneratorForStatements generator{m_context, m_utils};
+	for (VariableDeclaration const* variable: _contract.stateVariables())
+		if (!variable->isConstant() && !variable->immutable())
+			generator.initializeStateVar(*variable);
 
-			<assignToParams> <abiDecode>(memoryDataOffset, add(memoryDataOffset, argSize))
+	return generator.code();
+}
 
-			<constructorName>(<params>)
-		)X");
-		t("object", creationObjectName(_contract));
-		t("allocate", m_utils.allocationFunction());
-		t("assignToParams", paramVars == 0 ? "" : "let " + suffixedVariableNameList("param_", 0, paramVars) + " := ");
-		t("params", suffixedVariableNameList("param_", 0, paramVars));
-		t("abiDecode", abiFunctions.tupleDecoder(constructor->functionType(false)->parameterTypes(), true));
-		t("constructorName", m_context.enqueueFunctionForCodeGeneration(*constructor));
+void IRGenerator::generateImplicitConstructors(ContractDefinition const& _contract)
+{
+	auto listAllParams = [&](
+		map<ContractDefinition const*, string> const& baseParams,
+		ContractDefinition const* nextContract = nullptr) -> string
+	{
+		string paramsString;
+		if (nextContract != nullptr && baseParams.find(nextContract) != baseParams.end())
+			paramsString = baseParams.at(nextContract);
+		for (auto const& param: baseParams)
```

Comment:
This iterates in pointer order which can lead to non-deterministic compilation. Also I think that the parameters should always be listed in derived to base order.

Reply:
If we always iterate in derived to base order, then the special case about `nextContract` can be removed, can it?

Reply:
You can actually keep the map as it is, and just iterate using
`for (size_t i = 0; i < _contract.annotation().linearizedBaseContracts.size(); ++i)`

Reply:
I fixed like that